### PR TITLE
#75: Connect the current docker container with the ITDE docker network in bring_itde_up

### LINKED ISSUE #75: Connect the AI Lab docker container with the ITDE docker network in bring_itde_up
- If we start the ITDE from inside the AI Lab docker container, the AI Lab docker container is in a different docker network than the ITDE
- This means the AI Lab docker container can't access the DB
- To solve this, we need to connect the AI Lab Docker container to the network created by the ITDE
- However, because the ITDE is started from inside the AI Lab docker container, we don't know how the AI Lab container is named
- To find the correct container name for the AI Lab container, we need to look in through all containers and somehow identify the container we are running in
- After some research, we concluded we can use the IP addresses of the container
  - From inside the container, we can get the available interfaces and their IP addresses
  - From docker we can get with docker inspect the connected networks and the corresponding IP addresses
  - By matching these, we can find the correct container because the IP addresses need to be unique within a docker daemon

### PR DESCRIPTION
**Pull request by tkilias**
Fixes #75

### CODE REVIEW

File: exasol/itde_manager.py [hunk 1]
```diff
@@ -77,6 +82,14 @@ def bring_itde_up(conf: Secrets) -> None:
     conf.save(AILabConfig.cert_vld, "False")
 
 
+def _add_current_container_to_db_network(container_info: ContainerInfo):
+    network_name = container_info.network_info.network_name
+    docker_client = docker.from_env()
```
Comment (tomuben) [inline exasol/itde_manager.py:87]: don't forget to call `docker_client.close()` at the end. I remember that we had an issue in ITDE with too many unclosed docker clients....

File: test/unit/test_current_container_finder.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+class TestSetup:
+    def __init__(self, containers: List[Dict[str, str]], ip_addresses: List[str]):
+        self.docker_client_mock = create_docker_client_mock(containers)
+        self.ip_retriever_mock = create_ip_retriever_mock(ip_addresses)
+        self.current_container_finder = CurrentContainerFinder(
+            docker_client=self.docker_client_mock, ip_retriever=self.ip_retriever_mock
+        )
+
+
+def test_ips_in_ips_of_single_container():
```
Comment (ckunki) [inline test/unit/test_current_container_finder.py:71]: This test was quite hard to understand for me. 

Questions: 
1. For finding a specific container do the ip addresses of all its networks need to be included in the list of searched ip addresses?

```suggestion
def test_single_container():
    """
    Verify finding a single container with all ip addresses exposed by its networks being contained in the given whitelist.
    """
```

File: test/unit/test_current_container_finder.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    test_setup = TestSetup(
+        containers=[
+            {"test1": "192.168.0.1", "test2": "192.168.1.1"},
```
Comment (ckunki) [inline test/unit/test_current_container_finder.py:74]: Maybe we should name the containers?
* `matching_container`
* `non_matching_container`

Maybe also the edge case of a container with some but not all ips contained in the whitelist would be worth a test?

File: test/unit/test_current_container_finder.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    assert test_setup.docker_client_mock.mock_calls == [call.containers.list()]
+    assert test_setup.ip_retriever_mock.mock_calls == [call.ips()]
+    assert (
+        result
+        == mock_cast(test_setup.docker_client_mock.containers.list).return_value[0]
+    )
+
+
+def test_ips_not_in_container_ips():
```
Comment (ckunki) [inline test/unit/test_current_container_finder.py:90]:
```suggestion
def test_no_matching_container():
```

File: test/unit/test_current_container_finder.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    )
+
+    result = test_setup.current_container_finder.current_container()
+
+    assert test_setup.docker_client_mock.mock_calls == [call.containers.list()]
+    assert test_setup.ip_retriever_mock.mock_calls == [call.ips()]
+    assert result == None
+
+
+def test_ips_in_ips_of_multiple_container():
```
Comment (ckunki) [inline test/unit/test_current_container_finder.py:106]:
```suggestion
def test_multiple_containers_matching():
```

**Review by ckunki — CHANGES_REQUESTED** [review]

File: test/integration/test_itde_manager_in_docker_container.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def remove_itde():
+    remove_docker_container([DB_CONTAINER_NAME])
+    remove_docker_networks([DB_NETWORK_NAME])
+    remove_docker_volumes([DB_VOLUME_NAME])
+
+
+@pytest.fixture
+def dockerfile() -> str:
```
Comment (tomuben) [inline test/integration/test_itde_manager_in_docker_container.py:43]: I would call it `dockerfile_content`

File: exasol/itde_manager.py (hunk 1 above)
Comment (tomuben) [reply exasol/itde_manager.py:87]: I saw in the test you are already using `ContextDockerClient`....

File: test/integration/test_itde_manager_in_docker_container.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+from exasol_integration_test_docker_environment.lib.docker.volumes.utils import (
+    remove_docker_volumes,  # type: ignore
+)
+from exasol_integration_test_docker_environment.lib.test_environment.docker_container_copy import (
+    DockerContainerCopy,
+)
+
+from exasol.utils import upward_file_search
+
+TEST_CONTAINER = "itde_manager_test_container"
```
[the remark this answers is not in the document]
Comment (tkilias) [reply test/integration/test_itde_manager_in_docker_container.py:27]: I keep it by test_container, because manager is alone a non-descriptive word and in the end the container is for running tests

File: test/integration/test_itde_manager_in_docker_container.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    source_code = f"{function_source_code}\nrun_test()"
+    return source_code
+
+
+@pytest.fixture
+def docker_container(wheel_path, function_source_code, docker_image):
+    with ContextDockerClient() as client:
+        container = client.containers.run(
+            docker_image.id,
+            name=TEST_CONTAINER,
```
[the remark this answers is not in the document]
Comment (tkilias) [reply test/integration/test_itde_manager_in_docker_container.py:121]: we stay with test container

File: test/unit/test_current_container_finder.py (hunk 2 above)
Comment (tkilias) [reply test/unit/test_current_container_finder.py:71]: yes, we don't know which IP  will match, this method is hacky enough that we need to be prepared for everything.

File: test/integration/test_itde_manager_in_docker_container.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            )
+            assert exit_code == 0, output
+            yield container
+        finally:
+            remove_docker_container([container.id])
+
+
+def test_bring_itde_up(docker_container):
+    exec_result = docker_container.exec_run("python3 /tmp/test.py")
+    assert exec_result.exit_code == 0, exec_result.output
```
[the remark this answers is not in the document]
Comment (tkilias) [reply test/integration/test_itde_manager_in_docker_container.py:159]: The test script is already connecting to the DB and to be honest, if the ITDE doesn't run in its own container, I don't care, important is that we can connect to it from a container

However, the remove, I have simply forgot, that I need to add.

File: test/unit/test_current_container_finder.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    def __init__(self, containers: List[Dict[str, str]], ip_addresses: List[str]):
+        self.docker_client_mock = create_docker_client_mock(containers)
+        self.ip_retriever_mock = create_ip_retriever_mock(ip_addresses)
+        self.current_container_finder = CurrentContainerFinder(
+            docker_client=self.docker_client_mock, ip_retriever=self.ip_retriever_mock
+        )
+
+
+def test_ips_in_ips_of_single_container():
+    test_setup = TestSetup(
```
[the remark this answers is not in the document]
Comment (tkilias) [reply test/unit/test_current_container_finder.py:79]: I think, testee doesn't fit here perfectly, because the Testee would be for me the Finder, but only the finder

File: exasol/current_container_finder.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        if len(current_container_candidates) == 1:
+            return next(iter(current_container_candidates))
+        elif len(current_container_candidates) == 0:
+            return None
+        else:
+            raise RuntimeError(
+                f"Multiple potential current containers found: {current_container_candidates}"
+            )
+
+    def retrieve_networks_of_container(
```
[the remark this answers is not in the document]
Comment (tkilias) [reply exasol/current_container_finder.py:47]: agree, this is a left over from refactoring

File: exasol/current_container_finder.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+            for network in self.retrieve_networks_of_container(container)
+            if network["IPAddress"] in ip_addresses
+        }
+        if len(current_container_candidates) == 1:
+            return next(iter(current_container_candidates))
+        elif len(current_container_candidates) == 0:
+            return None
+        else:
+            raise RuntimeError(
+                f"Multiple potential current containers found: {current_container_candidates}"
```
[the remark this answers is not in the document]
Comment (tkilias) [reply exasol/current_container_finder.py:44]: maybe a bit overkill, but I can add it

File: exasol/itde_manager.py (hunk 1 above)
Comment (tkilias) [reply exasol/itde_manager.py:87]: good point

**Review by ckunki — CHANGES_REQUESTED** [review]

**Review by ckunki — DISMISSED** [review]

**Review by ckunki — DISMISSED** [review]

**Review by ckunki — APPROVED** [review]
